### PR TITLE
Rework gcc diagnostic color patch

### PR DESCRIPTION
**Pull request by Zapeth**
Reworked the patch based on method described here -> https://github.com/BurntSushi/ripgrep/issues/94#issuecomment-261761687
(its mentioned that `GetFileInformationByHandleEx` wont work for this in Windows XP as opposed to `NtQueryObject`, but I'm guessing thats not something to consider anymore anyway)

I'm not sure if the check for `\cygwin` is necessary, its mentioned in the above linked thread so I included it to be safe, but none of the ttys I tested resolved to that name, maybe its from old mintty versions?

Probably fixes https://github.com/msys2/MINGW-packages/issues/14087 and https://github.com/msys2/MINGW-packages/issues/8600, I tested it on my system (Win7 x64) and diagnostic output is coloured again by default.

### CODE REVIEW

**Comment by Zapeth** [comment]
Not sure why CI failed, error doesn't look like its related to my changes
```
   ../../../gcc-12.2.0/libcpp/system.h:404:30: error: expected identifier before string constant
    404 | #define abort() fancy_abort (__FILE__, __LINE__, __FUNCTION__)
        |                              ^~~~~~~~
  ../../../gcc-12.2.0/libcpp/system.h:404:30: error: expected ',' or '...' before string constant
  ../../../gcc-12.2.0/libcpp/system.h:404:30: error: expected identifier before string constant
    404 | #define abort() fancy_abort (__FILE__, __LINE__, __FUNCTION__)
        |                              ^~~~~~~~
  ../../../gcc-12.2.0/libcpp/system.h:404:30: error: expected ',' or '...' before string constant
  make[3]: *** [Makefile:227: files.o] Error 1
  checking for ANSI C header files... make[3]: Leaving directory '/C/_/mingw-w64-gcc/src/build-MINGW64/build-x86_64-w64-mingw32/libcpp'
  make[2]: *** [Makefile:3121: all-build-libcpp] Error 2
  make[2]: *** Waiting for unfinished jobs....
```

**Comment by Biswa96** [comment]
> Not sure why CI failed, error doesn't look like its related to my changes

Yes, the issue had been fixed in this upstream commit https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=902c755930326cb4405672aa3ea13c35c653cbff. Please make sure it can be applied in current gcc version.

**Comment by Zapeth** [comment]
Are you suggesting I should apply the linked commit as a new patch? Shouldn't that be done in a separate PR?

**Comment by Biswa96** [comment]
> Shouldn't that be done in a separate PR?

Agree, that would be better.

**Comment by lazka** [comment]
> Shouldn't that be done in a separate PR?

I'll look into that.

**Comment by lazka** [comment]
The backport doesn't help. Not sure.

**Comment by Biswa96** [comment]
lh_mouse shared a sed command here https://gcc.gnu.org/pipermail/gcc-patches/2023-January/609514.html. Can you try that?

**Comment by lazka** [comment]
thanks, I got help on IRC

File: mingw-w64-gcc/0140-gcc-diagnostic-color.patch [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ 
++  bool ret = false;
+   h = GetStdHandle (STD_ERROR_HANDLE);
+-  return (h != INVALID_HANDLE_VALUE) && (h != NULL)
++  ret = (h != INVALID_HANDLE_VALUE) && (h != NULL)
+ 	  && GetConsoleMode (h, &m);
++
++  if (!ret)
++    {
++      ULONG size = sizeof (FILE_NAME_INFO) + MAX_PATH;
```
Comment (lazka) [inline mingw-w64-gcc/0140-gcc-diagnostic-color.patch:17]:
```suggestion
+      ULONG size = sizeof (FILE_NAME_INFO) + (MAX_PATH * sizeof (wchar_t));
```

Comment (Zapeth) [reply mingw-w64-gcc/0140-gcc-diagnostic-color.patch:17]: Whats the preferred method of accepting suggestions, via web interface (separate commit) or rebase/force-push?

Comment (lazka) [reply mingw-w64-gcc/0140-gcc-diagnostic-color.patch:17]: someone needs to update the checksum anyway, so this isn't enough. this just seemed like an easy way to point out a potential change.

Whatever you prefer.

Comment (Zapeth) [reply mingw-w64-gcc/0140-gcc-diagnostic-color.patch]: Ah right, I forgot about having to update the checksum. Anway, thanks for the suggestion

**Pull request merged**